ompi/mca/coll/ucc/coll_ucc_ops.c

```diff
@@ -0,0 +1,249 @@
+}
+
+int mca_coll_ucc_allreduce(const void *sbuf, void *rbuf, int count, struct ompi_datatype_t *dtype,
+                           struct ompi_op_t *op, struct ompi_communicator_t *comm,
+                           mca_coll_base_module_t *module)
+{
+    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
+    ucc_coll_req_h         req;
+    ucc_datatype_t         ucc_dt;
+    ucc_reduction_op_t     ucc_op;
+
+    UCC_VERBOSE(3, "running ucc allreduce");
+    ucc_dt = ompi_dtype_to_ucc_dtype(dtype);
+        ucc_op = ompi_op_to_ucc_op(op);
+    if (OPAL_UNLIKELY(COLL_UCC_DT_UNSUPPORTED == ucc_dt ||
+                      COLL_UCC_OP_UNSUPPORTED == ucc_op)) {
+        UCC_VERBOSE(5, "ompi_datatype is not supported: dtype = %s; calling fallback allreduce;",
+                     dtype->super.name);
+        goto fallback;
+    }
+    ucc_coll_args_t coll = {
+        .mask = UCC_COLL_ARGS_FIELD_PREDEFINED_REDUCTIONS,
+        .coll_type = UCC_COLL_TYPE_ALLREDUCE,
+        .src.info = {
+            .buffer   = (void*)sbuf,
+            .count    = count,
+            .datatype = ucc_dt,
+            .mem_type = UCC_MEMORY_TYPE_UNKNOWN
+        },
+        .dst.info = {
+            .buffer   = rbuf,
+            .mem_type = UCC_MEMORY_TYPE_UNKNOWN
+        },
+        .reduce = {
+            .predefined_op = ucc_op,
+        },
+    };
+    if (MPI_IN_PLACE == sbuf) {
+        coll.mask |= UCC_COLL_ARGS_FIELD_FLAGS;
+        coll.flags = UCC_COLL_ARGS_FLAG_IN_PLACE;
+    }
+    COLL_UCC_CHECK(ucc_collective_init(&coll, &req, ucc_module->ucc_team));
+    COLL_UCC_CHECK(ucc_collective_post(req));
+    COLL_UCC_CHECK(coll_ucc_req_wait(req));
+    return OMPI_SUCCESS;
+fallback:
+    UCC_VERBOSE(3, "running fallback allreduce");
+    return ucc_module->previous_allreduce(sbuf, rbuf, count, dtype, op,
+                                          comm, ucc_module->previous_allreduce_module);
+}
+
+
+int mca_coll_ucc_alltoall(const void *sbuf, int scount, struct ompi_datatype_t *sdtype,
+                           void* rbuf, int rcount, struct ompi_datatype_t *rdtype,
+                           struct ompi_communicator_t *comm,
+                           mca_coll_base_module_t *module)
+{
+    mca_coll_ucc_module_t *ucc_module = (mca_coll_ucc_module_t*)module;
+    ucc_coll_req_h         req;
+    size_t                 dt_size;
+    ucc_datatype_t         ucc_sdt, ucc_rdt;
+    UCC_VERBOSE(3, "running ucc alltoall");
+
+    if (!ompi_datatype_is_contiguous_memory_layout(sdtype, scount) ||
+        !ompi_datatype_is_contiguous_memory_layout(rdtype, rcount)) {
+        goto fallback;
+    }
+    ucc_sdt = ompi_dtype_to_ucc_dtype(sdtype);
+    ucc_rdt = ompi_dtype_to_ucc_dtype(rdtype);
+    if (COLL_UCC_DT_UNSUPPORTED == ucc_sdt ||
+        COLL_UCC_DT_UNSUPPORTED == ucc_rdt) {
+        goto fallback;
+    }
+
+    opal_datatype_type_size(&sdtype->super, &dt_size);
+    ucc_coll_args_t coll = {
```

Comment:
here we definitely need to set mask to zero. If it contains garbage we can end up with inplace alltoall accidently